src/ring_theory/dedekind_domain.lean

```diff
@@ -108,33 +108,150 @@ structure is_dedekind_domain_dvr : Prop :=
+    have := fractional_ideal.mem_span_singleton_self (generator (I : submodule R₁ g.codomain)),
+    rw [hI, fractional_ideal.mem_zero_iff] at this,
+    contradiction }
+end
+
+lemma is_principal_inv (I : fractional_ideal g)
+  [submodule.is_principal (I : submodule R₁ g.codomain)] (h : I ≠ 0) :
+  submodule.is_principal (I⁻¹).1 :=
+begin
+  rw [fractional_ideal.val_eq_coe, fractional_ideal.is_principal_iff],
+  use (generator (I : submodule R₁ g.codomain))⁻¹,
+  have hI : I  * fractional_ideal.span_singleton ((generator (I : submodule R₁ g.codomain))⁻¹)  = 1,
+  apply @mul_generator_self_inv _ _ _ _ _ I _ h,
+  apply (@right_inverse_eq _ _ _ _ _ I (fractional_ideal.span_singleton ((generator (I : submodule R₁ g.codomain))⁻¹)) hI).symm,
+end
+
 /--
 A Dedekind domain is an integral domain that is not a field such that every fractional ideal has an inverse.
 
 This is equivalent to `is_dedekind_domain`.
 TODO: prove the equivalence.
 -/
+
+
 structure is_dedekind_domain_inv : Prop :=
 (not_is_field : ¬ is_field A)
-(mul_inv_cancel : ∀ I ≠ (⊥ : fractional_ideal (fraction_ring.of A)), I * I⁻¹ = 1)
-
-section
+(mul_inv_cancel : ∀ I ≠ (⊥ : fractional_ideal (fraction_ring.of A)), I * 1 / I = 1)
 
 open ring.fractional_ideal
 
-lemma is_dedekind_domain_inv_iff (f : fraction_map A K) :
-  is_dedekind_domain_inv A ↔
-    (¬ is_field A) ∧ (∀ I ≠ (⊥ : fractional_ideal f), I * I⁻¹ = 1) :=
-begin
-  split; rintros ⟨hf, hi⟩; use hf; intros I hI,
-  { have := hi (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom I) (map_ne_zero _ hI),
-    erw [← map_inv, ← fractional_ideal.map_mul] at this,
-    convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom) this;
-      simp only [alg_equiv.to_alg_hom_eq_coe, map_symm_map, map_one] },
-  { have := hi (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom I) (map_ne_zero _ hI),
-    erw [← map_inv, ← fractional_ideal.map_mul] at this,
-    convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom) this;
-      simp only [alg_equiv.to_alg_hom_eq_coe, map_map_symm, map_one] }
-end
+-- lemma is_dedekind_domain_inv_iff (f : fraction_map A K) :
+--   is_dedekind_domain_inv A ↔
+--     (¬ is_field A) ∧ (∀ I ≠ (⊥ : fractional_ideal f), I * I⁻¹ = 1) :=
+-- begin
+--   split; rintros ⟨hf, hi⟩; use hf; intros I hI,
+--   { have := hi (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom I) (map_ne_zero _ hI),
+--     erw [← map_inv, ← fractional_ideal.map_mul] at this,
+--     convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom) this;
+--       simp only [alg_equiv.to_alg_hom_eq_coe, map_symm_map, map_one] },
+--   { have := hi (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom I) (map_ne_zero _ hI),
+--     erw [← map_inv, ← fractional_ideal.map_mul] at this,
+--     convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom) this;
+--       simp only [alg_equiv.to_alg_hom_eq_coe, map_map_symm, map_one] }
+-- end
```

Comment:
This lemma should be kept (after we fix it), since it proves `is_dedekind_domain_inv` doesn't depend on the choice of `fraction_map`.
```suggestion
lemma is_dedekind_domain_inv_iff (f : fraction_map A K) :
  is_dedekind_domain_inv A ↔
    (¬ is_field A) ∧ (∀ I ≠ (⊥ : fractional_ideal f), I * I⁻¹ = 1) :=
begin
  split; rintros ⟨hf, hi⟩; use hf; intros I hI,
  { have := hi (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom I) (map_ne_zero _ hI),
    erw [← map_inv, ← fractional_ideal.map_mul] at this,
    convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom) this;
      simp only [alg_equiv.to_alg_hom_eq_coe, map_symm_map, map_one] },
  { have := hi (map (fraction_ring.alg_equiv_of_quotient f).to_alg_hom I) (map_ne_zero _ hI),
     erw [← map_inv, ← fractional_ideal.map_mul] at this,
     convert congr_arg (map (fraction_ring.alg_equiv_of_quotient f).symm.to_alg_hom) this;
       simp only [alg_equiv.to_alg_hom_eq_coe, map_map_symm, map_one] }
end
```

Reply:
Well, I agree: but the way it is it does not compile, so I thought about commenting (not deleting!) it for the PR in order to get all the `inv` stuff removed from `fractional_ideal`; and eventually de-commenting it in another smaller PR dedicated to resolving the proof. Doesn't it make sense?

Reply:
In principle it's good to keep the amount of functionality intact while we refactor, to prove that the refactor is sensible. I pushed a fixed version of the proof so we don't have to worry about this point anyway :)